### ISSUE TITLE
iviewer Shape propagation

### ISSUE THREAD

**Issue by will-moore** [openmicroscopy/design#88]
Major feature supported by Insight but not by iviewer 1.0.0.

See previous ideas at https://github.com/openmicroscopy/design/issues/59

Goals:
 - Allow Shape 'propagation' / 'duplication' within ROI as in Insight 
 - Want to avoid repeated propagation of shape to all indexes within ROI - resulting in many duplicates on the same plane (early versions of iviewer).
 - Would be nice to handle "Batch edit" of shapes as discussed with users recently at TIM Dusseldorf:
    - e.g. resize a bunch of Rectangle shapes on different planes.
    - Currently, we can batch-select and move shapes by dragging (even if on different Z-planes) but can't resize multiple shapes.
 - Limitation of Insight propagation is that the viewer doesn't show planes that you are propagating to.
 - Want to support multiple ROI shapes on the same Z/T plane (not supported in Insight) and allow different types of shape e.g. Rectangle and Ellipse in one ROI, as allowed by the server.

Suggestions:

 Here, we add a "ROI Grid" option under Edit (could also put this option under a right-click menu on ROI/Shape).

![roi-grid1](https://user-images.githubusercontent.com/900055/37972766-f0cd805c-31d0-11e8-8359-32d4c0a08e10.jpg)

This displays the shapes under ROI as in the Insight grid.

![roi-grid2](https://user-images.githubusercontent.com/900055/37972811-09f50fa0-31d1-11e8-86fd-7f5dbf3e6807.jpg)

 - The selected Z/T index is shown by the blue selected square. Selecting a Z/T index will show that plane in the viewer AND select the shape(s) in this ROI on that plane in the viewer.
 - Edit > Copy (or keyboard Ctrl C shortcut) will copy the Shape(s) on that plane.

![roi-grid3](https://user-images.githubusercontent.com/900055/37973046-8940a0d0-31d1-11e8-9a3e-0429d73d5735.jpg)

 - We can then select multiple planes by dragging as in Insight or by Shift-Click (range) or Ctrl-Click (toggle) and paste Shapes to multiple planes.

**Comment by will-moore** [comment]
cc @waxenegger @pwalczysko @jburel @mporter-gre Feedback & thoughts welcome!

**Comment by pwalczysko** [comment]
> If we ensure that a contiguous range is selected ...

Does that mean that the user can paste the ROI only on contiguous range of z/t ? I hope not.

**Comment by pwalczysko** [comment]
I would add an axis legend to the grid (z, t). It is hard to realize what the grid actually shows otherwise.

**Comment by pwalczysko** [comment]
``ROI grid`` - if the user does not know upfront what that is, there is no idea about why and how this could be useful. Wouldn't it be better to use the word ``propagate`` ?

**Comment by will-moore** [comment]
> Does that mean that the user can paste the ROI only on contiguous range of z/t ? I hope not.

Yes - this is the current behaviour in Insight - I can only propagate to a contiguous set of planes at a time. The ROI grid approach is actually more flexible since I can Copy and Paste to any planes, even if they are not contiguous with the original plane.

Adding an axis makes sense, agreed.

I chose "ROI grid" because it is really just a different layout and is not only for propagation. You could simply use it as a different way of seeing the distribution of Shapes in an ROI. Actually, it might be useful to enable the same view of ALL shapes on the image, not just a single ROI. This would need more thought about how to handle display, copy and paste etc, but the view itself could be handy.

The propagation happens simply because you are able to select more than 1 plane to paste the shapes to (and we also know to put all the shapes in the same ROI).
However, it could also be nice to allow propagation without ROI grid by copying a shape, selecting a range of planes (using Z-projection) and pasting to all those planes. The only question in this case would be "Do we put all these shapes in the same ROI or separate ROIs?".

**Comment by will-moore** [comment]
@pwalczysko It may be that "ROI grid" is not the best term (especially since it will be a single "row" or "column" for most images). I just don't want to pick the wrong name simply to align with Insight behaviour, but it might be that "propagate" is actually the best in this case.

**Comment by jburel** [comment]
The term "grid" leads to something different for me, a feature requested by user, the ability to "browse" an image by plane, we could misuse the multiview for that and put each plane in a grid.
That's for  a different scoping issue
We should avoid the term grid for propagation

**Comment by will-moore** [comment]
After more discussion with @pwalczysko and @jburel, I think this is diverging into two independent features:
 - Paste to multiple planes (probably selecting multiple planes via Z/T projection, although JM has some other ideas)
 - Grid-view of Shapes - This would look like the github activity heatmap and provide a useful overview of where your shapes are in the image. Could also be used to select Z/T planes and ranges.

I'll describe these latest ideas on a new issue, to avoid confusion with those above.
